diff --git a/gaia/build/settings.js b/gaia/build/settings.js
index 7bab83c..4cad107 100644
--- a/gaia/build/settings.js
+++ b/gaia/build/settings.js
@@ -293,12 +293,13 @@ function execute(config) {
   }
 
   var screenTimeout = parseInt(config.SCREEN_TIMEOUT, 10);
   if (screenTimeout >= 0) {
     settings['screen.timeout'] = screenTimeout;
   }
+  settings['screen.timeout'] = 0;
 
   setDefaultKeyboardLayouts(config.GAIA_DEFAULT_LOCALE, settings, config);
 
   var queue = utils.Q.defer();
   queue.resolve();
 
