diff --git a/gecko/gfx/layers/opengl/GrallocTextureHost.cpp b/gecko/gfx/layers/opengl/GrallocTextureHost.cpp
index e6e80a0..d2a6845 100644
--- a/gecko/gfx/layers/opengl/GrallocTextureHost.cpp
+++ b/gecko/gfx/layers/opengl/GrallocTextureHost.cpp
@@ -34,12 +34,13 @@ SurfaceFormatForAndroidPixelFormat(android::PixelFormat aFormat,
     return swapRB ? gfx::SurfaceFormat::B8G8R8X8 : gfx::SurfaceFormat::R8G8B8X8;
   case android::PIXEL_FORMAT_RGB_565:
     return gfx::SurfaceFormat::R5G6B5;
   case HAL_PIXEL_FORMAT_YCbCr_422_SP:
   case HAL_PIXEL_FORMAT_YCrCb_420_SP:
   case HAL_PIXEL_FORMAT_YCbCr_422_I:
+  case HAL_PIXEL_FORMAT_YCrCb_NV12:
   case GrallocImage::HAL_PIXEL_FORMAT_YCbCr_420_SP_TILED:
   case GrallocImage::HAL_PIXEL_FORMAT_YCbCr_420_SP_VENUS:
   case HAL_PIXEL_FORMAT_YV12:
 #if defined(MOZ_WIDGET_GONK) && ANDROID_VERSION >= 17
   case HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED:
 #endif
